src/btree/bt_read.c

```diff
@@ -152,6 +152,78 @@ __instantiate_birthmarks(WT_SESSION_IMPL *session, WT_REF *ref)
     return (ret);
 }
 
+/*
+ * __wt_modify_vector_init --
+ *     Initialise a modify vector.
+ */
+void
+__wt_modify_vector_init(WT_MODIFY_VECTOR *modifies, WT_SESSION_IMPL *session)
+{
+    WT_CLEAR(*modifies);
+    modifies->session = session;
+    modifies->listp = modifies->list;
+}
+
+/*
+ * __wt_modify_vector_push --
+ *     Push a modify update pointer to a modify vector. If we exceed the allowed stack space in the
+ *     vector, we'll be doing malloc here.
+ */
+int
+__wt_modify_vector_push(WT_MODIFY_VECTOR *modifies, WT_UPDATE *upd)
+{
+    WT_DECL_RET;
+    bool migrate_from_stack;
+
+    migrate_from_stack = false;
+
+    if (modifies->size >= WT_MODIFY_VECTOR_STACK_SIZE) {
+        if (modifies->capacity == 0 && modifies->size == WT_MODIFY_VECTOR_STACK_SIZE) {
+            migrate_from_stack = true;
+            modifies->listp = NULL;
+        }
+        WT_ERR(__wt_realloc_def(
+          modifies->session, &modifies->capacity, modifies->size + 1, &modifies->listp));
+        if (migrate_from_stack)
+            memcpy(modifies->listp, modifies->list, sizeof(modifies->list));
+    }
+    modifies->listp[modifies->size++] = upd;
+    return (0);
+
+err:
+    if (modifies->listp == NULL) {
+        modifies->listp = modifies->list;
+        modifies->capacity = 0;
```

Comment:
I'm unsure what you mean. If we jump to `err` it means we failed to allocate. I just set `allocated_bytes` to 0 to ensure that we don't think we have heap memory to free when we call `__wt_modify_vector_free`. I don't think we need to do anything with `size` since we wouldn't have incremented it yet. The struct should look the way it did before the `push` call.

That was what I was aiming for when I wrote this. Does something seem fishy here?

Reply:
> I don't think we need to do anything with size since we wouldn't have incremented it yet. 

But on the second & subsequent calls to realloc, won't size be non-zero? IOW, the 2nd time we call realloc, it fails.

Reply:
96f3eeb

I still don't see the problem. I've added a comment explaining what I'm trying to do. If there's still an issue here let me know and I can fix it in a subsequent change.